varianteval/scripts/annotation_script.py

```diff
@@ -0,0 +1,124 @@
+"""
+Script to add annotation to input callset with respect to specified
+comparison callsets
+"""
+
+import sys
+import re
+import numpy as np
+from pysam import VariantFile
+import argparse
+from collections import defaultdict
+
+
+def overlapping_records(intrvl_A, intrvl_B, overlap_threshold=0.5, bp_margin=5000):
+    '''
+    Helper method to determine whether two records' start-end
+    intervals overlap to a given breakpoint margin, and whether
+    the intervals' overlap is >=0.5*(svlen) for interval A and B
+    '''
+    bp_check = (np.abs(intrvl_A[0] - intrvl_B[0]) < bp_margin \
+                and np.abs(intrvl_A[1] - intrvl_B[1]) < bp_margin)
+    intrvl_overlap = (min(intrvl_A[1], intrvl_B[1])) - (max(intrvl_A[0], intrvl_B[0]))
+    overlap_check = (intrvl_overlap >= overlap_threshold * (intrvl_A[1] - intrvl_A[0])) and \
+                    (intrvl_overlap >= overlap_threshold * (intrvl_B[1] - intrvl_B[0]))
+
+    return bp_check and overlap_check
+
+
+def build_callset_dict(callset_vcf):
+    record_dict = defaultdict(list)
+    callset = VariantFile(callset_vcf)
+    for rec in callset.fetch():
+        record_dict[rec.chrom].append(
+            ((rec.start, rec.stop), rec.info['SVTYPE']))
+    return record_dict
+
+
+def annotate_source_vcf(source_vcf, annotated_vcf_dir, comparison_vcfs, comparison_labels, overlap_threshold=0.5,
+                        bp_margin=5000, type_constraint='agnostic'):
+    """
+    Inputs
+        source_vcf: vcf that be augmented with annotations
+        annotated_vcf_dir: output directory for annotated vcf
+        comparison_vcfs: list of vcfs to be checked for overlap with records in source_vcf
+        comparison_labels: list of labels to be used in annotation (corresponding to comparison_vcfs list)
+        overlap_threshold: threshold for labeled two records as overlapping
+    Output
+        Annotated vcf placed in the specified directory with records' INFO fields amended with labels indicating
+        overlap with the input comparison vcfs
+    """
+    assert len(comparison_vcfs) == len(comparison_labels), \
+        'comparison_vcfs and comparison_labels do not have matching length'
+
+    vcf_in_file = VariantFile(source_vcf)
+    header = vcf_in_file.header
+    # DB may be redundant with the vcf-specific labels, but also gives a compact field for just showing
+    # the callset collisions without the extra information about the conflicting record
+    if 'DB' not in header.info:
+        header.info.add('DB', number=1, type='String', description="Databases also containing this record")
```

Comment:
nit: rename "DB" and "ANNOT" to reflect what they actually collect; potentially callset collisions could also be represented as a binary vector (like what survivor does)

Reply:
'DB' -> 'OVERLAP'
'ANNOT' -> 'OVERLAPINFO'